trace-cruncher: Fix CI issue and add reproducibility tools

knauth opened the pull request:
This commit adds a scripts/ directory and the date-snapshot.sh script.
date-snapshot automates the process of checking out the last commit
in a repository before a specific date.

We leverage this tool in the GitHub actions CI to fix the ongoing build
error by locking our build dependencies to a specific date- currently
June 1, 2022. This date can be advanced when the breaking change is
found and fixed in the dependencies.

This commit is the beginning of work to make the trace-cruncher build
fully reproducible for containerization.

Signed-off-by: June Knauth (VMware) <june.knauth@gmail.com>

knauth committed:
trace-cruncher: Fix CI issue and add reproducibility tools

This commit adds a scripts/ directory and the date-snapshot.sh script.
date-snapshot automates the process of checking out the last commit
in a repository before a specific date.

We leverage this tool in the GitHub actions CI to fix the ongoing build
error by locking our build dependencies to a specific date- currently
June 1, 2022. This date can be advanced when the breaking change is
found and fixed in the dependencies.

This commit is the beginning of work to make the trace-cruncher build
fully reproducible for containerization.

Signed-off-by: June Knauth (VMware) <june.knauth@gmail.com>

tzstoyanov commented on scripts/date-snapshot/date-snapshot.sh:
```diff
@@ -0,0 +1,65 @@
+#!/bin/bash
+
```
All trace-cruncher source files should have a license identifier and a copyright notice, at the beginning of the file. For scripts it should be right after the command interpreter. When you make important modifications to the file, you can also add your name to the top of the file. Please, add here:
```
# SPDX-License-Identifier: LGPL-2.1
# Copyright (C) 2022, VMware Inc, June Knauth <june.knauth@gmail.com>

```

tzstoyanov reviewed (changes requested):
Hi June,
Thanks for working on improving the trace-cruncher CI. The changes that you proposed look good to me in general, I have only a few comments.
The description of the commit is really very good, but I have a suggestion on how to write even better commit messages. There is no need to explain what code or files are changed, as this is visible in the changes. Instead, focus more on the invisible things - what the changes do, the motivation of the changes, what problem they fix, why they should be accepted (you did that very well in your commit message!).

tzstoyanov commented on scripts/date-snapshot/repos:
```diff
@@ -0,0 +1,4 @@
+libtraceevent;https://git.kernel.org/pub/scm/libs/libtrace/libtraceevent.git/;libtraceevent;20220601
+libtracefs;https://git.kernel.org/pub/scm/libs/libtrace/libtracefs.git/;libtracefs;20220601
+trace-cmd;https://git.kernel.org/pub/scm/utils/trace-cmd/trace-cmd.git/;master;20220601
+kernel-shark;https://github.com/yordan-karadzhov/kernel-shark-v2.beta.git/;kernelshark;20220601
```
The `date-snapshot.sh` script generates directories, using the names from this file. All these auto generated directories should be added into the `.gitignore` file.
Also, these directories should be deleted by `make clean`. Hint: you can enhance the `date-snapshot.sh` script with an additional parameter to clean all auto generated files / directories and to trigger that logic from `make clean`.

tzstoyanov commented on scripts/date-snapshot/date-snapshot.sh:
```diff
@@ -0,0 +1,65 @@
+  IN="${1}"
+  # $fields=(${IN//;/ }) 
+  IFS=';' read -ra ADDR <<< "$IN"
+  
```
Add a check if the destination directory is empty, or may be just delete it if already exists to ensure `git clone` will not be confused.

warthog9 replied:
I'd argue not stomping on the directory is more better?  Might need an option of "--preclean" or something to do deletes if it finds something it doesn't like or expect

yordan-karadzhov commented on scripts/date-snapshot/repos:
```diff
@@ -0,0 +1,4 @@
+libtraceevent;https://git.kernel.org/pub/scm/libs/libtrace/libtraceevent.git/;libtraceevent;20220601
+libtracefs;https://git.kernel.org/pub/scm/libs/libtrace/libtracefs.git/;libtracefs;20220601
+trace-cmd;https://git.kernel.org/pub/scm/utils/trace-cmd/trace-cmd.git/;master;20220601
+kernel-shark;https://github.com/yordan-karadzhov/kernel-shark-v2.beta.git/;kernelshark;20220601
```
This is not the correct kernelshark repo, I wonder how you even managed to find this one.

Use ths one on kernel.org:
git://git.kernel.org/pub/scm/utils/trace-cmd/kernel-shark.git

knauth replied:
This was the source specified in the original GitHub actions script, so I copied it over. Thanks for clarifying that we should be using the kernel.org version.

knauth committed:
trace-cruncher: Implement PR feedback

This commit adds arguments to the date-snapshot script to enable deleting existing directories and either exiting or proceeding from there. This functionality is used in `make clean` to remove build artifacts. It also eases iterative testing when using the script.

In addition, the git source for kernel-shark has been changed to the mainline kernel.org repo because the old remote was used for testing and was not intended for stable release.

Finally, whitespace was cleaned up and invisible characters removed for easier readability.

Signed-off-by: June Knauth (VMware) <june.knauth@gmail.com>

knauth commented:
@warthog9 @tzstoyanov @yordan-karadzhov Just pushed a commit with the requested changes, please review.

warthog9 commented on scripts/date-snapshot/date-snapshot.sh:
```diff
@@ -1,44 +1,74 @@
+    rm -rf ./${ADDR[0]}
+  fi
+
+  # if download is true and directory doesn't exist
```
I'd add a quick warning on if [[ ! -d ${ADDR[0]} ]];do echo "*** Warning: a non directory where we are expecting a directory - nothing will be done there";done

would be my suggestion so that's warned appropriately in case someone does something really weird

warthog9 commented on scripts/date-snapshot/date-snapshot.sh:
```diff
@@ -1,44 +1,74 @@
+    cd ..
+  elif [[ $download = false ]]; then
+    echo "Download flag not set, skipping ${ADDR[0]}"
+  elif [[ -d ${ADDR[0]} ]] ; then
```
Anything with a file path needs "'s around it, always in bash

/I am A Directory that will break your checking/

for example

warthog9 commented on scripts/date-snapshot/date-snapshot.sh:
```diff
@@ -49,14 +79,14 @@ while test $# -gt 0; do
       ;;
     *)
-      echo "Usage: ${package} [-s] [-i|--input] [-f|--file]"
+      echo "Usage: ${package} [-c|--clean] [-d|delete] [-i|--input] [-f|--file] <input>"
```
adding -h/--help and have it spit out your help text is pretty normal too (this is a nicety not required)

knauth committed:
trace-cruncher: Refactor and improve date-snapshot

This commit improves the behavior of date-snapshot in unexpected conditions and improves readability.

Explicit conditionals added as they are best practice to ensure consistent behavior.
In addition, quotes added around all inputs to the script to prevent multiple arguments from being passed to commands.
Finally, added a warning when a directory that is about to be deleted does not exist so that users are warned in unexpected situations.

Signed-off-by: June Knauth (VMware) <june.knauth@gmail.com>

tzstoyanov commented on scripts/date-snapshot/date-snapshot.sh:
```diff
@@ -1,44 +1,74 @@
-  cd .. 
+  if $clean ; then # delete the repos if requested
+    echo "Removing ${ADDR[0]}"
+    rm -rf ./${ADDR[0]}
```
This logic assumes that the script is executed within the `scripts/date-snapshot` directory. That is not the case when the script is started from the top Makefile, then it is executed from the project's top directory and the logic fails.  You can get the path from `${0}` of the script and append it here, instead of that `./`. Or even better - look what is passed as `-f` argument and use that path, the directories should be there.

tzstoyanov commented on scripts/date-snapshot/date-snapshot.sh:
```diff
@@ -1,44 +1,74 @@
-   
-  cd .. 
+  if $clean ; then # delete the repos if requested
+    echo "Removing ${ADDR[0]}"
```
There should be no logs on the console if there is no error. You can add new `-v` argument, to set the desired log level and filter all informational and warning logs depending of the desired verbosity. By default, only critical logs should be printed.

tzstoyanov commented on scripts/date-snapshot/date-snapshot.sh:
```diff
@@ -1,44 +1,74 @@
+  # if download is true and directory doesn't exist
+  if [[ $download = true && ! -d ${ADDR[0]} ]] ; then
+    git clone -b ${ADDR[2]} ${ADDR[1]} ${ADDR[0]}
+    cd ${ADDR[0]}
```
The same comment as above - the correct path should be appended to `${ADDR[0]}`,  as the script may be executed from another directory.

tzstoyanov commented on scripts/date-snapshot/date-snapshot.sh:
```diff
@@ -1,44 +1,74 @@
+    # this checks out the first entry in the list of commit hashes which occur before our date
+    git checkout `git log --date=unix --before=${last_date} --pretty=format:"%H" | head -n 1`
+
+    cd ..
```
Better approach is to save the original directory and use it here, instead of that relative path.

tzstoyanov commented on scripts/date-snapshot/date-snapshot.sh:
```diff
@@ -1,44 +1,74 @@
+  elif [[ -d ${ADDR[0]} ]] ; then
+    echo "Directory exists, skipping ${ADDR[0]}"
+  fi
+
```
All these logs should not be printed by default, they should be filtered based on the logs verbosity level.

tzstoyanov reviewed (changes requested):
Thanks June!
I have a few more comments.

tzstoyanov commented on scripts/date-snapshot/date-snapshot.sh:
```diff
@@ -7,35 +7,39 @@
+    else
+      echo "Removing ${ADDR[0]}"
+      rm -rf "./${ADDR[0]}"
+    fi
```
That log should not be displayed by default.

tzstoyanov commented on scripts/date-snapshot/date-snapshot.sh:
```diff
@@ -7,35 +7,39 @@
-    rm -rf ./${ADDR[0]}
+  if [[ $clean = "true" ]] ; then # delete the repos if requested
+    if [[ ! -d "${ADDR[0]}" ]] ; then
+      echo "*** Warning: Directory ${ADDR[0]} does not exist, nothing to delete"
```
That is not a critical error and should not be displayed by default.

knauth committed:
trace-cruncher: Add more flags and workdir tracking to date-snapshot

This commit further extends the feature set of date-snapshot and helps
ensure it will behave as expected in all situations.

Added logging levels so that users can debug the behavior of the script.
By default, only critical errors are printed.

Added a --latest flag to bypass the specified date and get the latest version
of the repository. This will be leveraged in GitHub actions to check for
regressions or breaking changes automatically.

Added workdir tracking to the script. When the -f flag is passed, the workdir
is the parent dir of the provided file. When -i is passed, the workdir is the
current working directory.

Workdir tracking ensures commands like -d will always execute in the same
directory as the repos were created in, as long as the input file is the same.

Signed-off-by: June Knauth (VMware) <june.knauth@gmail.com>

tzstoyanov reviewed (approved):
Looks good to me, thanks June!
